### ISSUE TITLE
Astro fails to output a hoisted .js file in dist

### ISSUE THREAD

**Issue by kamoshi** [withastro/astro#7146]
### What version of `astro` are you using?

2.5.0

### Are you using an SSR adapter? If so, which one?

None

### What package manager are you using?

pnpm

### What operating system are you using?

Kubuntu 23.04

### What browser are you using?

Firefox

### Describe the Bug

Given a file `src/utils/test.ts` in `src`:
```ts
export function bindSubtitle() {
  const target = document.getElementById("subtitle");
  target && (target.innerText = "hello");
}
```

When I try to import and use it in an `.astro` file like this:
```html
<body>
	<slot />
	<script>
		import { bindSubtitle } from "../utils/test";

		bindSubtitle();
	</script>
</body>
```

The resulting bundle contains a script which tries to import JS code from a nonexistent file, e.g `import"./hoisted.8cca6103.js"`, which effectively breaks many minified scripts across the website.
![image](https://github.com/withastro/astro/assets/18511281/06abc221-a7b7-4fe8-9fbc-72a0e5eaadad)

This problem only occurs in the build output, not when running dev mode. I've provided a stackblitz link with an example configuration that results in a broken build.

### Link to Minimal Reproducible Example

https://stackblitz.com/edit/github-ksnujv?file=dist%2F_astro%2Fhoisted.97534249.js

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by edoardofelici** [comment]
The same is happening to me, even though I fixed all astro related packages versions to avoid these very types of problems. Now my lazy loaded images don't load at all because the hoisted file that astro is ignoring is the one containing some code related to the lazy loading. 

It's pretty critical I'd say. 

I had everything fixed down at astro@2.2.3. Is astro not following semver? 

![astro-hoisted](https://github.com/withastro/astro/assets/25923450/c9695908-999c-4e83-85d0-bd0f036282ea)

**Comment by edoardofelici** [comment]
Another note, I think the missing hoisted code is related to <Component client:visible />. 

Stuff that should be loaded on 'client:visible' directive is also not working, which makes me think that the required code is in that hoisted file. 

Hope this helps narrow things down.

**Comment by srflp** [comment]
bumping rollup via resolutions in my package.json:
```
  "resolutions": {
    "rollup": "3.22.0"
  }
```
causes the same problem

**Comment by kamoshi** [comment]
It seems it might be caused by this particular pull request in Rollup

https://github.com/rollup/rollup/pull/4989

**Comment by kamoshi** [comment]
This seems to be fixed in 2.5.2, so closing 🙂

**Comment by isimmons** [comment]
I have this issue in latest astro 4 which is using rollup version 4.9.2, astro 4.10 and astro/vercel integration 6.1.2

Failed to load resource: the server responded with a status of 404 () hoisted.j4pK5_uY.js:1

It didn't seem to affect anything until I tried adding a svelte component and now I also get 
Uncaught (in promise) TypeError: Failed to fetch dynamically imported module) ComponentName.dlfhdlfjlsl.js

It's a static import but I guess the dynamic part comes from the way Astro loads it because of the client:load directive?
